### PR TITLE
Part 1: Bring KVIrc to the 21st century

### PR DESCRIPTION
**Pull request by staticfox**
Avoid many un-necessary memory allocations as well as start the migration to unique pointers where appropriate. Start migrating out of the KVirc standard library into the C++ standard library.

---

I tested this and it all seems to work, let's see what AV and Travis have to say. Note, I'm doing this in chunks to preserve  1) my own sanity 2) anyone who views the commit's sanity.

EDIT:
As a side note, `KviPtrListIterator` is very temporary as that whole system (`KviPointerList` and co.) are going away.

### CODE REVIEW

File: src/kvirc/ui/KviChannelWindow.cpp [hunk 1]
```diff
@@ -1680,15 +1667,15 @@ void KviChannelWindow::getChannelActivityStats(KviChannelActivityStats * pStats)
 	tTwoMinsAgo -= 120;
 	tNow -= 60;
 
-	KviPointerHashTable<QString, int> userDict;
-	userDict.setAutoDelete(false);
+	std::set<QString> userDict;
 
-	int iFake;
```
Comment (un1versal) [inline src/kvirc/ui/KviChannelWindow.cpp:1686]: good riddens on iFake

reviewing this with Github-Dark is better ;)

**Comment by un1versal** [comment]
Well
- [x] 1) bug copy and past something into input line then move cursor left o right or delete even
  demo-a

**Note** if you paste the contents are correct

**Note2**
If you copy two words only and paste them they are pasted first word second and second word first 3 words and things go silly
- [x] 2) bug  stats while connecting to server

capture
- [x] 3) IP of my own connection via `/stats l` 255.255.255.255 **nice**

[11:32:55] [barjavel.freenode.net][211] un1versal un1versal[~error@255.255.255.255] 0 355 43 31 0 :123 0 -

Im pretty sure there's more but thats something to start
- [x] 4) outpust of `/stats p` is also useless

```
[11:42:12] u p 0
[11:42:12] End of /STATS report
```

File: .travis.yml [hunk 2]
```diff
@@ -51,14 +51,15 @@ install:
           # want cmake newer than by default in travis osx
           brew upgrade cmake
           export CMAKE_FRAMEWORK_PATH="$CMAKE_FRAMEWORK_PATH:$(brew --prefix qt5)"
+          ADDITIONAL_LINK_ARGS="-stdlib=libc++ -mmacosx-version-min=10.7"
```
Comment (DarthGandalf) [inline .travis.yml]: I think this should be fixed not here, but in CMakeLists.txt.
Everyone should be able to build on OS X, not only Travis.

**Comment by un1versal** [comment]
- [x] 5) typing stuff is all backwards and crazy demo-a
- [x] 6) tab complete prefix vs nicname also backwards
  demo-a
- [ ] 7) testing bug #1309 as requested, saved a theme with icons, switched to it, shortly KVIrc crashed...#

seems it crashes here
`
0x00000000005b3343 in QTreeWidgetItem::setExpanded (this=0x0, aexpand=true) at /usr/include/x86_64-linux-gnu/qt5/QtWidgets/qtreewidget.h:399
399 { if (view) view->setItemExpanded(this, aexpand); }
`
- [x] 8) Editing topic its all backwards
  demo-a

File: src/kvirc/kernel/KviIrcConnection.h [hunk 3]
```diff
@@ -160,8 +161,8 @@ class KVIRC_API KviIrcConnection : public QObject
 
 	KviIrcConnectionStateData * m_pStateData; // owned, never null
 
-	KviPointerList<KviChannelWindow> * m_pChannelList; // owned, never null, elements shallow
-	KviPointerList<KviQueryWindow> * m_pQueryList;     // owned, never null, elements shallow
+	std::vector<KviChannelWindow *> m_pChannelList; // owned, never null, elements shallow
+	std::vector<KviQueryWindow *> m_pQueryList;     // owned, never null, elements shallow
```
Comment (DarthGandalf) [inline src/kvirc/kernel/KviIrcConnection.h]: This looks wrong...
Either the comment is outdated, and it's not owned anymore, or the windows are owned, then they should be stored as `vector<unique_ptr<KQW>>`

File: src/kvirc/kernel/KviNotifyList.h [hunk 4]
```diff
@@ -76,18 +78,18 @@ class KVIRC_API KviIsOnNotifyListManager : public KviNotifyListManager
 	~KviIsOnNotifyListManager();
 
 private:
-	KviPointerHashTable<QString, QString> * m_pRegUserDict; // dict notifystring->reguser name
-	KviPointerList<QString> * m_pNotifyList;                // list of notifystring (total)
-	KviPointerList<QString> * m_pIsOnList;                  // list of notifystring (one session)
+	std::map<QString, std::unique_ptr<QString>> m_pRegUserDict; // dict notifystring->reguser name
+	std::vector<std::unique_ptr<QString>> m_NotifyList;     // list of notifystring (total)
```
Comment (DarthGandalf) [inline src/kvirc/kernel/KviNotifyList.h]: Why not just `vector<QString>`?

**Comment by staticfox** [comment]
see if 1, 5, 6, and 8 are fixed (also can you turn them in to checkboxes to make it easier to track? :)

**Comment by un1versal** [comment]
> see if 1, 5, 6, and 8 are fixed (also can you turn them in to checkboxes to make it easier to track? :)
- [x]  checkboxes check

will test soon enough and tick whats fixed

**edit**

1, 5, 6 and 8 been ticked off (on) cant duplicate that anymore.

the copy and pasting the first time around wasn't immediate theres was a 1 second lag between pushing **ctrl+v** and the actual paste appearing. (not sure if its relevant but cant remember ever experiencing any lag)

**Comment by staticfox** [comment]
@un1versal test 2, 3, and 4 now?

File: .travis.yml [hunk 5]
```diff
@@ -51,14 +51,15 @@ install:
...
         fi
 script:
     - |
         if [ -z "$DEBUILD" ]
             then
             mkdir build
             cd build
-            cmake $CMAKE_ARGS .. || exit 1
+            cmake $CMAKE_ARGS -DADDITIONAL_LINK_FLAGS="$ADDITIONAL_LINK_ARGS" .. || exit 1
```
Comment (un1versal) [inline .travis.yml]: you probably going to do it eventually but here is a reminder anywho. squash this with https://github.com/kvirc/KVIrc/pull/2079/commits/23bf81d1d26c3ac3370f1fdec7e0e5a71607b0bd

**Comment by un1versal** [comment]
It would be extra nice in the end if we could avoid another clang type situation that made blame impossible to load.

**Comment by un1versal** [comment]
3 is still broken. 2 and 4 been ticked fixed

**edit**

I would check on your side Linux what the output of all /stats requests is. here I get really borked results still
capture

**Comment by staticfox** [comment]
thats how 3 normally looks though

**Comment by Dessa** [comment]
/stats l looks fine on a server im operd on. it seems the stats l output is some ircd-seven thing.

**Comment by un1versal** [comment]
> /stats l looks fine on a server im operd on. it seems the stats l output is some ircd-seven thing.

best check Linux side stats m here output garbage mostly, but since Im not opped everywhere and ones Im opped on block stats reports cant really test properly.

> thats how 3 normally looks though

never noticed it like that before. with a 255.255.255.255 seems wrong?

**Comment by staticfox** [comment]
> never noticed it like that before. with a 255.255.255.255 seems wrong?

it's right

**Comment by un1versal** [comment]
> it's right

if you say so, (ticked box)

**Comment by un1versal** [comment]
re: nr 7 

**edit**

changing themes kills kvirc no matter if saving theme before or changing to any global same result as below

```
(gdb) attach 4086
Attaching to process 4086
[New LWP 4089]
[New LWP 4092]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f3bbb538e8d in poll () at ../sysdeps/unix/syscall-template.S:84
84  ../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) bt full
#0  0x00007f3bbb538e8d in poll () at ../sysdeps/unix/syscall-template.S:84
No locals.
#1  0x00007f3bb981239c in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#2  0x00007f3bb98124ac in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#3  0x00007f3bbc384a7f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
No symbol table info available.
#4  0x00007f3bbc32bdea in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
No symbol table info available.
#5  0x00007f3bbc333e8c in QCoreApplication::exec() () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
---Type <return> to continue, or q <return> to quit---
No symbol table info available.
#6  0x00000000006f9e7b in main (argc=2, argv=0x7ffc462a1478) at /home/un1versal/Desktop/kvirc/src/kvirc/kernel/KviMain.cpp:488
        a = {argc = 2, argv = 0x7ffc462a1478, configFile = 0x0, createFile = false, bForceNewSession = false, bShowPopup = true, bExecuteCommandAndClose = false, 
          szExecCommand = {static null = {<No data fields>}, d = 0x7f3bbc4016e0 <QArrayData::shared_null>}, szExecRemoteCommand = {static null = {<No data fields>}, 
            d = 0x7f3bbc4016e0 <QArrayData::shared_null>}}
        iRetCode = 0
        pTheApp = 0x115d280
        szRemoteCommand = {static null = {<No data fields>}, d = 0x7f3bbc4016e0 <QArrayData::shared_null>}
        __PRETTY_FUNCTION__ = "int main(int, char**)"
        iRetVal = 91137
(gdb) c
Continuing.

Thread 1 "kvirc" received signal SIGSEGV, Segmentation fault.
0x00000000005ad10b in QTreeWidgetItem::setExpanded (this=0x0, aexpand=true) at /usr/include/x86_64-linux-gnu/qt5/QtWidgets/qtreewidget.h:399
399 { if (view) view->setItemExpanded(this, aexpand); }
(gdb) c
Continuing.
[Thread 0x7f3bad6ae700 (LWP 4092) exited]
[Thread 0x7f3baf906700 (LWP 4089) exited]

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) 
```

**Comment by staticfox** [comment]
@un1versal I tested #7 on mac, windows, arch, ubuntu 16.04 and 14.04 and it doesn't crash so idk

**Comment by un1versal** [comment]
good for you I guess bad news for me cause it crashes here, (well unless it was fixed in last minutes by some change)

Also thanks for merging this without considering blame or any of the comments I made

File: src/kvirc/kernel/KviApplication.cpp [hunk 6]
```diff
@@ -1103,39 +1097,29 @@ void KviApplication::setAvatarOnFileReceived(
     const QString & szUser,
     const QString & szHost)
 {
-	if(!m_pPendingAvatarChanges)
+	if(m_PendingAvatarChanges.size() >= KVI_MAX_PENDING_AVATARS) // can't be...
 	{
-		m_pPendingAvatarChanges = new KviPointerList<KviPendingAvatarChange>;
-		m_pPendingAvatarChanges->setAutoDelete(true);
+		m_PendingAvatarChanges.erase(m_PendingAvatarChanges.begin()); // kill the first entry
```
Comment (DarthGandalf) [inline src/kvirc/kernel/KviApplication.cpp:1102]: It's a random entry, not the first one.
unordered_map doesn't have an order.

**Pull request merged**